### PR TITLE
Improving hints generation to avoid using the same commitment twice

### PR DESCRIPTION
**Pull request by kushti**
In previous implementation of distributed signatures, it was possible for the prover to generate a single commitment for Bob for the statement "atLeast(3, Coll(pubkeyA, pubkeyB, pubkeyC, pubkeyD, pubkeyE)) && (pubkeyB || pubkeyF)", however, a signature done for the single commitment is not secure (allows for efficient extraction of Bob's private key).

In this PR hints made tied with positions of the nodes they are related to. And for the example statement from the above, two commitments are to be generated for Bob. 
Hints matching is fixed accordingly.

### CODE REVIEW

File: sigmastate/src/main/scala/sigmastate/trees.scala [hunk 1]
```diff
@@ -16,13 +16,29 @@ import sigmastate.utxo.{Transformer, SimpleTransformerCompanion}
 import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
 
+
+/**
+  * Basic trait for inner nodes of crypto-trees, so AND/OR/THRESHOLD sigma-protocol connectives
+  */
+trait SigmaConjecture extends SigmaBoolean {
+  val children: Seq[SigmaBoolean]
```
Comment (aslesarenko) [inline sigmastate/src/main/scala/sigmastate/trees.scala:24]: Let's avoid using vals in new traits (legacy code aside) and use defs instead.

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala [hunk 2]
```diff
@@ -1,66 +1,121 @@
...
+            val hints = Seq(OwnCommitment(leaf, r, a, position), RealCommitment(leaf, a, position))
+            bag.addHints(hints: _*)
+          } else {
+            bag
+          }
+      }
+    }
+
+    traverseNode(sigmaTree, HintsBag.empty, position = Hint.CryptoTreePrefix)
+  }
```
Comment (aslesarenko) [inline sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala:62]: Please add test-vector-like tests for this method, especially boundary cases and negative case, which are not covered now. 
This is to catch regressions, because applications will rely on exact behavior of this code.

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala [hunk 3]
```diff
@@ -156,19 +177,38 @@ trait ProverInterpreter extends Interpreter with ProverUtils with AttributionCor
...
+  private def setPositions(uc: UnprovenConjecture): UnprovenConjecture = {
+    val updChildren = uc.children.zipWithIndex.map { case (pt, idx) =>
+        pt.asInstanceOf[UnprovenTree].withPosition(uc.position + "-" + idx.toString)
+    }
+    uc match {
+      case and: CAndUnproven => and.copy(children = updChildren)
+      case or: COrUnproven => or.copy(children = updChildren)
+      case threshold: CThresholdUnproven => threshold.copy(children = updChildren)
+    }
+  }
```
Comment (aslesarenko) [inline sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala:192]: Please add test-vector-like tests for this method, this will also help see how the positions are assigned.
This is to catch regressions, because applications will rely on exact behavior of this code.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala (hunk 3 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala:192]: Please also base this PR on top of #677

File: sigmastate/src/main/scala/sigmastate/interpreter/Hint.scala [hunk 4]
```diff
@@ -12,7 +12,45 @@ import sigmastate.basics.VerifierMessage.Challenge
   * and the prover knows only a secret for the public key pk1, the prover fails on proving without a hint. But if the
   * prover knows that pk2 is known to another party, the prover may prove the statement (with an empty proof for "pk2").
   */
-trait Hint
+trait Hint {
+
+  /**
+    * A hint is related to a subtree (or a leaf) of a tree.
+    *
+    * Position is encoded like (the example below is for CTHRESHOLD(2, Seq(pk1, pk2, pk3 && pk4)) :
```
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/Hint.scala:20]: @aslesarenko  Like the tree below obviously

File: sigmastate/src/main/scala/sigmastate/UnprovenTree.scala [hunk 5]
```diff
@@ -35,56 +33,82 @@ trait ProofTreeConjecture extends ProofTree {
...
+/**
+  * A node of a sigma-tree used by the prover
+  */
 sealed trait UnprovenTree extends ProofTree {
+
+  /**
+    * Positon of the node in the tree, see comments for `position` field in
+    * `sigmastate.interpreter.Hint`
+    */
+  val position: String
```
Comment (greenhat) [inline sigmastate/src/main/scala/sigmastate/UnprovenTree.scala:45]: Could some ADT or newtype be a better fit then String type here?

**Review by greenhat — CHANGES_REQUESTED** [review]

File: sigmastate/src/main/scala/sigmastate/UnprovenTree.scala [hunk 6]
```diff
@@ -35,56 +33,82 @@ trait ProofTreeConjecture extends ProofTree {
...
+
+  /**
+    * Node's sigma-protocol statement to be proven.
+    */
   val proposition: SigmaBoolean
 
+  /**
+    * Whether the node represents simulated sigma-protocol
+    */
```
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/UnprovenTree.scala:54]: Added  a reference to the trait's ScalaDoc.

File: sigmastate/src/main/scala/sigmastate/UnprovenTree.scala [hunk 7]
```diff
@@ -35,56 +33,82 @@ trait ProofTreeConjecture extends ProofTree {
...
+  override def withPosition(updatedPosition: String): UnprovenTree = this.copy(position = updatedPosition)
 }
 
 case class CThresholdUnproven(override val proposition: CTHRESHOLD,
                        override val challengeOpt: Option[Challenge] = None,
                        override val simulated: Boolean,
                        k: Integer,
                        children: Seq[ProofTree],
-                       polynomialOpt: Option[GF2_192_Poly]) extends UnprovenConjecture {
+                       polynomialOpt: Option[GF2_192_Poly],
```
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/UnprovenTree.scala:110]: Added reference to a paper.

File: sigmastate/src/main/scala/sigmastate/UnprovenTree.scala [hunk 8]
```diff
@@ -35,56 +33,82 @@ trait ProofTreeConjecture extends ProofTree {
...
+    * Whether the node represents simulated sigma-protocol
+    */
   val simulated: Boolean
 
   def real: Boolean = !simulated
 
+  /**
+    * Challenge used by the prover.
+    */
   val challengeOpt: Option[Array[Byte]]
```
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/UnprovenTree.scala:62]: Trait comment now contains a link to the whitepaper section which is describing sigma protocols in brief and provides links for further reading.

File: sigmastate/src/main/scala/sigmastate/interpreter/Hint.scala [hunk 9]
```diff
@@ -12,7 +12,45 @@ import sigmastate.basics.VerifierMessage.Challenge
...
+    *           /   |
+    *         0-2-0 0-2-1
+    *
+    *  So a hint associated with pk1 has a position "0-0", pk4 - "0-2-1" .
+    *
+    *  Please note that "0" prefix is for a crypto tree. There are several kinds of trees during evaluation.
+    *  Initial mixed tree (ergoTree) would have another prefix.
+    *
+    *
+    *
```
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/Hint.scala:38]: fixed

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala [hunk 10]
```diff
@@ -1,66 +1,121 @@
...
+                             context: CTX,
+                             generateFor: Seq[SigmaBoolean]): HintsBag = {
+    val reducedTree = fullReduction(ergoTree, context, Interpreter.emptyEnv)._1
+    generateCommitmentsFor(reducedTree, generateFor)
+  }
+
+  /**
+    * A method which is is generating commitments for all the public keys provided.
+    *
+    * Currently only ProveDlog and ProveDiffieHellman are supported.
```
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala:27]: comment improved

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala [hunk 11]
```diff
@@ -1,66 +1,121 @@
...
+    *
+    * @param sigmaTree - crypto-tree
+    * @param generateFor - public keys
```
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala:30]: done

File: sigmastate/src/main/scala/sigmastate/trees.scala (hunk 1 above)
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/trees.scala:24]: done

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala (hunk 3 above)
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala:191]: rebased

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala (hunk 2 above)
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala:62]: test done, see ProverSpecification

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala (hunk 3 above)
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala:191]: test done , see ProverSpecification

File: sigmastate/src/main/scala/sigmastate/UnprovenTree.scala (hunk 5 above)
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/UnprovenTree.scala]: Made NodePosition ADT

**Comment by kushti** [comment]
@aslesarenko @greenhat comments addresses, please make another pass!

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala [hunk 12]
```diff
@@ -30,6 +30,27 @@ trait ProverInterpreter extends Interpreter with ProverUtils with AttributionCor
 
   val secrets: Seq[SigmaProtocolPrivateInput[_, _]]
```
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala:31]: Made public keys def instead.

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala (hunk 3 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala:192]: The method is not covered by the test
image

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala (hunk 2 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala:62]: DHT case is not covered
image

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala [hunk 13]
```diff
@@ -134,14 +128,14 @@ class DistributedSigSpecification extends SigmaTestingCommons {
     val env = Map("pubkeyA" -> pubkeyAlice, "pubkeyB" -> pubkeyBob, "pubkeyC" -> pubkeyCarol)
     val prop = compile(env, """atLeast(2, Coll(pubkeyA, pubkeyB, pubkeyC))""").asSigmaProp
 
-    val (rBob, aBob) = DLogInteractiveProver.firstMessage()
-    val dlBKnown: Hint = RealCommitment(pubkeyBob, aBob)
+    val bobHints = proverB.generateCommitments(prop.treeWithSegregation, ctx)
```
Comment (aslesarenko) [inline sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:131]: Where "She then puts Bob's randomness and fill his response with zero bits. Thus Alice's signature is not valid. Alice is sending her signature to Bob response with zero bits." mentioned in the description is done?

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala [hunk 14]
```diff
@@ -219,22 +213,22 @@ class DistributedSigSpecification extends SigmaTestingCommons {
...
 
     val bagC = proverC.bagForMultisig(ctx, prop, proofAlice.proof, Seq(pubkeyAlice), Seq(pubkeyDave)) ++
-      HintsBag(Seq(dlBKnown, OwnCommitment(pubkeyCarol, rCarol, aCarol)))
+      HintsBag(Seq(dlBKnown, carolHints.ownCommitments.head))
     val proofCarol = proverC.prove(prop, ctx, fakeMessage, bagC).get
 
     val bagB = (proverB.bagForMultisig(ctx, prop, proofAlice.proof, Seq(pubkeyAlice), Seq(pubkeyDave)) ++
                 proverB.bagForMultisig(ctx, prop, proofCarol.proof, Seq(pubkeyCarol)))
-                  .addHint(OwnCommitment(pubkeyBob, rBob, aBob))
+                  .addHint(bobHints.ownCommitments.head)
```
Comment (aslesarenko) [inline sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:231]: If I got it right, Carol's proof should also fail verification? Please add the check.

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala [hunk 15]
```diff
@@ -309,23 +303,24 @@ class DistributedSigSpecification extends SigmaTestingCommons {
     val script = """atLeast(4, Coll(pubkeyA, pubkeyB, pubkeyC, pubkeyD, pubkeyE, pubkeyF, pubkeyG, pubkeyH))"""
     val prop = compile(env, script).asSigmaProp
 
-    //Alice, Bob, Gerard, and Hannah are signing, others are simulated
+    // Alice, Bob, Gerard, and Hannah are signing, others are simulated
```
Comment (aslesarenko) [inline sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:306]: This is opportunity to improve diversity. Suggest to use at least one Asian, Black and pronoun name in the list.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala [hunk 16]
```diff
@@ -41,15 +39,13 @@ class DistributedSigSpecification extends SigmaTestingCommons {
     val env = Map("pubkeyA" -> pubkeyAlice, "pubkeyB" -> pubkeyBob)
     val prop: Values.Value[SSigmaProp.type] = compile(env, """pubkeyA && pubkeyB""").asSigmaProp
 
-    val (rBob, aBob) = DLogInteractiveProver.firstMessage()
-
-    val hintFromBob: Hint = RealCommitment(pubkeyBob, aBob)
-    val bagA = HintsBag(Seq(hintFromBob))
+    val hintsFromBob: HintsBag = proverB.generateCommitments(prop.treeWithSegregation, ctx)
```
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:42]: Are we going to cover tests with comments as well?

File: sigmastate/src/main/scala/sigmastate/interpreter/Hint.scala [hunk 17]
```diff
@@ -71,23 +80,28 @@ abstract class CommitmentHint extends Hint {
...
+                          override val position: NodePosition) extends CommitmentHint
 
 /**
   * A hint which contains a commitment to randomness associated with a public image of a secret.
   *
   * @param image      - image of a secret
   * @param commitment - commitment to randomness used while proving knowledge of the secret
   */
-case class SimulatedCommitment(override val image: SigmaBoolean, commitment: FirstProverMessage) extends CommitmentHint
+case class SimulatedCommitment(override val image: SigmaBoolean,
```
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/Hint.scala:102]: The type is encoding a boolean flag essentially, real or simulated

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala (hunk 15 above)
Comment (kushti) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:306]: Not willing to do that.

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:42]: @aslesarenko hints are not about commitments just, but also other parts of sigma protocol transcripts, and potentially not only.

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala (hunk 3 above)
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/ProverInterpreter.scala:192]: made new tests for this

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:42]: We're not necessarily covering tests by comments by an agreement.

hints is no synonymous to commitments. In general, hints could cover not just sigma protocol transcripts (they're about that now though).

File: sigmastate/src/main/scala/sigmastate/interpreter/Hint.scala (hunk 17 above)
[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/Hint.scala:102]: Real or simulated. Like a boolean flag, but encoded in type.

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala (hunk 13 above)
Comment (kushti) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:131]: In the prover

Comment (kushti) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:131]: In the prover

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala (hunk 14 above)
Comment (kushti) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:231]: done

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala (hunk 15 above)
Comment (kushti) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:306]: Do not understand you here. Not willing to do renaming (standard for all the cryptography textbooks and papers) also.

**Comment by kushti** [comment]
@aslesarenko improved test coverage for both the methods, addressed comment, please make another pass!

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:42]: As always, it is better to put the answer in the code. It is not useful here.

[the remark this answers is not in the document]
Comment (kushti) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:42]: @aslesarenko the answer in the code (if the question is about correspondence between hints and commitment) exists, that is, the HInt hierarchy.

File: sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala (hunk 2 above)
Comment (kushti) [reply sigmastate/src/main/scala/sigmastate/interpreter/ProverUtils.scala:62]: done

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:42]: This is conceptually not easy to understand scenario.
Isn't it better to have at least one such scenario fully documented?

**Review by aslesarenko — APPROVED** [review]
LGTM, except the DistributedSigSpecification is not readable enough.

File: sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala (hunk 13 above)
Comment (aslesarenko) [reply sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala:131]: It is also better to put as comments in the code.

**Comment by kushti** [comment]
> LGTM, except the DistributedSigSpecification is not readable enough.

@aslesarenko  Added reference to EIP11. For cases with simplified signing, see 2-out-of-2 and 2-out-of-3 comments. I guess this is enough, especially consider that this is the test.

**Review by greenhat — APPROVED** [review]
